metagenomics.py

```diff
@@ -721,63 +739,69 @@ def parser_krona(parser=argparse.ArgumentParser()):
     return parser
 
 
-def diamond(inBam, db, taxDb, outReport, outM8=None, outLca=None, numThreads=1):
+def diamond(inBam, db, taxDb, outReport, outReads=None, numThreads=1):
     '''
         Classify reads by the taxon of the Lowest Common Ancestor (LCA)
     '''
-    tmp_fastq = util.file.mkstempfname('.fastq')
-    tmp_fastq2 = util.file.mkstempfname('.fastq')
     # do not convert this to samtools bam2fq unless we can figure out how to replicate
     # the clipping functionality of Picard SamToFastq
     picard = tools.picard.SamToFastqTool()
-    picard_opts = {
-        'CLIPPING_ATTRIBUTE': tools.picard.SamToFastqTool.illumina_clipping_attribute,
-        'CLIPPING_ACTION': 'X'
-    }
-    picard.execute(
+    s2fq = picard.execute(
         inBam,
-        tmp_fastq,
-        tmp_fastq2,
-        picardOptions=tools.picard.PicardTools.dict_to_picard_opts(picard_opts),
-        JVMmemory=picard.jvmMemDefault
+        '/dev/stdout',
+        interleave=True,
+        illuminaClipping=True,
+        JVMmemory=picard.jvmMemDefault,
+        background=True,
+        stdout=subprocess.PIPE,
     )
 
     diamond_tool = tools.diamond.Diamond()
     diamond_tool.install()
-    tmp_alignment = util.file.mkstempfname('.daa')
-    tmp_m8 = util.file.mkstempfname('.diamond.m8')
-    diamond_tool.blastx(db, [tmp_fastq, tmp_fastq2], tmp_alignment, options={'--threads': numThreads})
-    diamond_tool.view(tmp_alignment, tmp_m8, options={'--threads': numThreads})
-
-    if outM8:
-        with open(tmp_m8, 'rb') as f_in:
-            with gzip.open(outM8, 'wb') as f_out:
-                shutil.copyfileobj(f_in, f_out)
+    taxonmap = join(taxDb, 'accession2taxid', 'prot.accession2taxid.gz')
+    taxonnodes = join(taxDb, 'nodes.dmp')
 
-    tax_db = TaxonomyDb(tax_dir=taxDb, load_names=True, load_nodes=True, load_gis=True)
-    tmp_lca_tsv = util.file.mkstempfname('.tsv')
-    with open(tmp_m8) as m8, open(tmp_lca_tsv, 'w') as lca:
-        blast_lca(tax_db, m8, lca, paired=True, min_bit_score=50)
+    cmd = '{} blastx --outfmt 102 --sallseqid'.format(diamond_tool.install_and_get_path())
+    if numThreads is not None:
+        cmd += ' --threads {threads}'.format(threads=numThreads)
+    cmd += ' --db {db} --taxonmap {taxonmap} --taxonnodes {taxonnodes}'.format(
+        threads=numThreads,
+        db=db,
+        taxonmap=taxonmap,
+        taxonnodes=taxonnodes)
 
-    if outLca:
-        with open(tmp_lca_tsv, 'rb') as f_in:
-            with gzip.open(outLca, 'wb') as f_out:
-                shutil.copyfileobj(f_in, f_out)
+    if outReads is not None:
+        cmd += ' | tee >(gzip > {out})'.format(out=outReads)
```

Comment:
Nice pipefitting. For our future selves, it may be worth adding a brief comment at the top describing the pipe setup in/out for this function.